InterruptClient.rb

```diff
@@ -0,0 +1,231 @@
+require 'socket'
+require 'json'
+
+require_relative './Console'
+
+class InterruptClient
+
+	CMD_QUIT = "\\"
+	START_MSG = 'Welcome! What is your name?'
+	INSTRUCTIONS = "Start typing to join the chat! To quit anytime, type #{CMD_QUIT}"
+	FAREWELL = "\r\nbye"
+	PROMPT = '> '
+
+	MAX_MSG_LENGTH = 3000 # max length of incoming message read
+	HANDSHAKE_WAIT = 2 # number of seconds to wait for ack from server before resending
+
+	COLORS = [
+		'green',
+		'magenta',
+		'cyan',
+		'blue',
+		'light_green',
+	]
+
+	def initialize(server_host, server_port)
+		@server_host = server_host
+		@server_port = server_port
+
+		@client = UDPSocket.new
+		@client.connect(@server_host, @server_port)
+
+		@latest_chat = '' # string timestamp of latest chat received
+	end
+
+	def run
+		set_name
+		handshake
+		instructions
+		terminal_config
+		receive_loop
+	end
+
+	def set_name
+		puts START_MSG
+		print PROMPT
+		@name = gets.chomp
+	end
+
+	def handshake
+		ackd = false
+		while not ackd
+			send_msg(msg_connect(@name))
+			incoming = IO.select([@client], nil, nil, HANDSHAKE_WAIT)
+			if ((not incoming.nil?) && ack?)
+				ackd = true
+			end
+		end
+	end
+
+	def ack?
+		data, sender = @client.recvfrom(MAX_MSG_LENGTH)
+		msg = parse_msg(data, sender)
+		return false if msg.nil?
+
+		msg['type'] == 'ack'
+	end
+
+	def instructions
+		print Console.clear
+		puts INSTRUCTIONS + "\n" * 12
+	end
+
+	def receive_loop
+		loop do
+			to_read = IO.select([@client, STDIN]) # check what's ready to read: socket, or terminal input
+			to_read[0].each { |ios|
+				if ios == @client
+					msg, sender = ios.recvfrom(MAX_MSG_LENGTH)
+					handle_msg(msg, sender)
+				elsif ios.tty? # ios comes from terminal
+					input = STDIN.getc
+					handle_key(input)
+				end
+			}
+		end
+	end
+
+	def handle_msg(msg, sender)
+		msg = parse_msg(msg, sender)
+		return if msg.nil?
+
+		case msg['type']
+		when 'chat'
+			time = msg['time']
+			return if time < @latest_chat
+
+			@latest_chat = time
+			chat_array = msg['body']
+			names_array = msg['names']
+
+			chat_string = chat_string(chat_array)
```

Comment:
Is this creating a new variable that shadows the function of the same name? I suggest using different names.